File: tdz/src/chunking.rs
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use chrono::{DateTime, Utc};
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum ChunkingLevel {
    Project,
    Module,
    Class,
    Method,
    Block,
}
impl ChunkingLevel {
    pub fn max_tokens(&self) -> usize {
        match self {
            ChunkingLevel::Project => 100,
            ChunkingLevel::Module => 500,
            ChunkingLevel::Class => 1000,
            ChunkingLevel::Method => 300,
            ChunkingLevel::Block => 100,
        }
    }
    pub fn description(&self) -> &'static str {
        match self {
            ChunkingLevel::Project => "High-level project planning and architecture",
            ChunkingLevel::Module => "Major system components and interfaces",
            ChunkingLevel::Class => "Class definitions and major functions",
            ChunkingLevel::Method => "Individual methods and helper functions",
            ChunkingLevel::Block => "Small code blocks and error handling",
        }
    }
    pub fn example(&self) -> &'static str {
        match self {
            ChunkingLevel::Project => "Build web scraper with database storage",
            ChunkingLevel::Module => "Create database handler module",
            ChunkingLevel::Class => "Implement DatabaseConnection class",
            ChunkingLevel::Method => "Write insert_record method",
            ChunkingLevel::Block => "Add error handling for connection timeout",
        }
    }
}
impl std::fmt::Display for ChunkingLevel {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ChunkingLevel::Project => write!(f, "project"),
            ChunkingLevel::Module => write!(f, "module"),
            ChunkingLevel::Class => write!(f, "class"),
            ChunkingLevel::Method => write!(f, "method"),
            ChunkingLevel::Block => write!(f, "block"),
        }
    }
}

impl std::str::FromStr for ChunkingLevel {
    type Err = crate::error::TodoziError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "project" => Ok(ChunkingLevel::Project),
            "module" => Ok(ChunkingLevel::Module),
            "class" => Ok(ChunkingLevel::Class),
            "method" => Ok(ChunkingLevel::Method),
            "block" => Ok(ChunkingLevel::Block),
            _ => Err(crate::error::TodoziError::ValidationError {
                message: format!("Invalid chunking level: {}", s),
            }),
        }
    }
}
impl std::fmt::Display for ChunkStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ChunkStatus::Pending => write!(f, "pending"),
            ChunkStatus::InProgress => write!(f, "in_progress"),
            ChunkStatus::Completed => write!(f, "completed"),
            ChunkStatus::Validated => write!(f, "validated"),
            ChunkStatus::Failed => write!(f, "failed"),
        }
    }
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectState {
    pub total_lines: usize,
    pub max_lines: usize,
    pub current_module: String,
    pub dependencies: Vec<String>,
    pub completed_modules: Vec<String>,
    pub pending_modules: Vec<String>,
    pub global_variables: HashMap<String, String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}
impl ProjectState {
    pub fn new(max_lines: usize) -> Self {
        let now = Utc::now();
        Self {
            total_lines: 0,
            max_lines,
            current_module: String::new(),
            dependencies: Vec::new(),
            completed_modules: Vec::new(),
            pending_modules: Vec::new(),
            global_variables: HashMap::new(),
            created_at: now,
            updated_at: now,
        }
    }
    pub fn to_state_string(&self) -> String {
        format!(
            r#"<project_state>
- Total lines written: {}/{}
- Current module: {}
- Dependencies: {}
- Completed modules: {}
- Pending modules: {}
- Global variables: {}
- Created: {}
- Updated: {}
</project_state>"#,
            self.total_lines, self.max_lines, self.current_module, self.dependencies
            .join(", "), self.completed_modules.join(", "), self.pending_modules
            .join(", "), self.global_variables.iter().map(| (k, v) | format!("{}={}", k,
            v)).collect::< Vec < _ >> ().join(", "), self.created_at
            .format("%Y-%m-%d %H:%M:%S"), self.updated_at.format("%Y-%m-%d %H:%M:%S")
        )
    }
    pub fn add_completed_module(&mut self, module: String) {
        if !self.completed_modules.contains(&module) {
            self.completed_modules.push(module);
            self.updated_at = Utc::now();
        }
    }
    pub fn add_pending_module(&mut self, module: String) {
        if !self.pending_modules.contains(&module) {
            self.pending_modules.push(module);
            self.updated_at = Utc::now();
        }
    }
    pub fn set_global_variable(&mut self, key: String, value: String) {
        self.global_variables.insert(key, value);
        self.updated_at = Utc::now();
    }
    pub fn increment_lines(&mut self, lines: usize) {
        self.total_lines += lines;
        self.updated_at = Utc::now();
    }
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContextWindow {
    pub previous_class: String,
    pub current_class: String,
    pub next_planned: String,
    pub global_vars_in_scope: Vec<String>,
    pub imports_used: Vec<String>,
    pub function_signatures: HashMap<String, String>,
    pub error_patterns_seen: Vec<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}
impl ContextWindow {
    pub fn new() -> Self {
        let now = Utc::now();
        Self {
            previous_class: String::new(),
            current_class: String::new(),
            next_planned: String::new(),
            global_vars_in_scope: Vec::new(),
            imports_used: Vec::new(),
            function_signatures: HashMap::new(),
            error_patterns_seen: Vec::new(),
            created_at: now,
            updated_at: now,
        }
    }
    pub fn to_context_string(&self) -> String {
        format!(
            r#"<context_window>
- Previous class: {}
- Current class: {}
- Next planned: {}
- Global variables in scope: {}
- Imports used: {}
- Function signatures: {}
- Error patterns seen: {}
- Created: {}
- Updated: {}
</context_window>"#,
            self.previous_class, self.current_class, self.next_planned, self
            .global_vars_in_scope.join(", "), self.imports_used.join(", "), self
            .function_signatures.iter().map(| (k, v) | format!("{}: {}", k, v))
            .collect::< Vec < _ >> ().join(", "), self.error_patterns_seen.join(", "),
            self.created_at.format("%Y-%m-%d %H:%M:%S"), self.updated_at
            .format("%Y-%m-%d %H:%M:%S")
        )
    }
    pub fn add_import(&mut self, import: String) {
        if !self.imports_used.contains(&import) {
            self.imports_used.push(import);
            self.updated_at = Utc::now();
        }
    }
    pub fn add_function_signature(&mut self, name: String, signature: String) {
        self.function_signatures.insert(name, signature);
        self.updated_at = Utc::now();
    }
    pub fn add_error_pattern(&mut self, pattern: String) {
        if !self.error_patterns_seen.contains(&pattern) {
            self.error_patterns_seen.push(pattern);
            self.updated_at = Utc::now();
        }
    }
    pub fn set_current_class(&mut self, class: String) {
        self.previous_class = self.current_class.clone();
        self.current_class = class;
        self.updated_at = Utc::now();
    }
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CodeChunk {
    pub chunk_id: String,
    pub status: ChunkStatus,
    pub dependencies: Vec<String>,
    pub code: String,
    pub tests: String,
    pub validated: bool,
    pub level: ChunkingLevel,
    pub estimated_tokens: usize,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum ChunkStatus {
    Pending,
    InProgress,
    Completed,
    Validated,
    Failed,
}
impl CodeChunk {
    pub fn new(chunk_id: String, level: ChunkingLevel) -> Self {
        let now = Utc::now();
        Self {
            chunk_id,
            status: ChunkStatus::Pending,
            dependencies: Vec::new(),
            code: String::new(),
            tests: String::new(),
            validated: false,
            level,
            estimated_tokens: 0,
            created_at: now,
            updated_at: now,
        }
    }
    pub fn add_dependency(&mut self, dep: String) {
        if !self.dependencies.contains(&dep) {
            self.dependencies.push(dep);
            self.updated_at = Utc::now();
        }
    }
    pub fn set_code(&mut self, code: String) {
        self.estimated_tokens = code.split_whitespace().count();
        self.code = code;
        self.updated_at = Utc::now();
    }
    pub fn set_tests(&mut self, tests: String) {
        self.tests = tests;
        self.updated_at = Utc::now();
    }
    pub fn mark_completed(&mut self) {
        self.status = ChunkStatus::Completed;
        self.updated_at = Utc::now();
    }
    pub fn mark_validated(&mut self) {
        self.validated = true;
        self.status = ChunkStatus::Validated;
        self.updated_at = Utc::now();
    }
    pub fn mark_failed(&mut self) {
        self.status = ChunkStatus::Failed;
        self.updated_at = Utc::now();
    }
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CodeGenerationGraph {
    pub chunks: HashMap<String, CodeChunk>,
    pub project_state: ProjectState,
    pub context_window: ContextWindow,
}
impl CodeGenerationGraph {
    pub fn new(max_lines: usize) -> Self {
        Self {
            chunks: HashMap::new(),
            project_state: ProjectState::new(max_lines),
            context_window: ContextWindow::new(),
        }
    }
    pub fn add_chunk(
        &mut self,
        chunk_id: String,
        level: ChunkingLevel,
        deps: Vec<String>,
    ) {
        let mut chunk = CodeChunk::new(chunk_id.clone(), level);
        for dep in deps {
            chunk.add_dependency(dep);
        }
        self.chunks.insert(chunk_id, chunk);
    }
    pub fn get_ready_chunks(&self) -> Vec<String> {
        let mut ready = Vec::new();
        for (chunk_id, chunk) in &self.chunks {
            if chunk.status == ChunkStatus::Pending {
                let deps_satisfied = chunk
                    .dependencies
                    .iter()
                    .all(|dep| {
                        self.chunks
                            .get(dep)
                            .map(|c| {
                                c.status == ChunkStatus::Completed
                                    || c.status == ChunkStatus::Validated
                            })
                            .unwrap_or(false)
                    });
                if deps_satisfied {
                    ready.push(chunk_id.clone());
                }
            }
        }
        ready
    }
    pub fn get_chunk(&self, chunk_id: &str) -> Option<&CodeChunk> {
        self.chunks.get(chunk_id)
    }
    pub fn get_chunk_mut(&mut self, chunk_id: &str) -> Option<&mut CodeChunk> {
        self.chunks.get_mut(chunk_id)
    }
    pub fn update_chunk_code(
        &mut self,
        chunk_id: &str,
        code: String,
    ) -> Result<(), String> {
        if let Some(chunk) = self.chunks.get_mut(chunk_id) {
            chunk.set_code(code);
            self.project_state.increment_lines(chunk.code.lines().count());
            Ok(())
        } else {
            Err(format!("Chunk {} not found", chunk_id))
        }
    }
    pub fn update_chunk_tests(
        &mut self,
        chunk_id: &str,
        tests: String,
    ) -> Result<(), String> {
        if let Some(chunk) = self.chunks.get_mut(chunk_id) {
            chunk.set_tests(tests);
            Ok(())
        } else {
            Err(format!("Chunk {} not found", chunk_id))
        }
    }
    pub fn mark_chunk_completed(&mut self, chunk_id: &str) -> Result<(), String> {
        if let Some(chunk) = self.chunks.get_mut(chunk_id) {
            chunk.mark_completed();
            self.project_state.add_completed_module(chunk_id.to_string());
            Ok(())
        } else {
            Err(format!("Chunk {} not found", chunk_id))
        }
    }
    pub fn mark_chunk_validated(&mut self, chunk_id: &str) -> Result<(), String> {
        if let Some(chunk) = self.chunks.get_mut(chunk_id) {
            chunk.mark_validated();
            Ok(())
        } else {
            Err(format!("Chunk {} not found", chunk_id))
        }
    }
    pub fn get_project_summary(&self) -> String {
        let completed_count = self
            .chunks
            .values()
            .filter(|c| {
                c.status == ChunkStatus::Completed || c.status == ChunkStatus::Validated
            })
            .count();
        let total_count = self.chunks.len();
        let pending_count = self
            .chunks
            .values()
            .filter(|c| c.status == ChunkStatus::Pending)
            .count();
        let in_progress_count = self
            .chunks
            .values()
            .filter(|c| c.status == ChunkStatus::InProgress)
            .count();
        format!(
            r#"<project_summary>
- Total chunks: {}
- Completed: {}
- In progress: {}
- Pending: {}
- Project state: {}
- Context window: {}
</project_summary>"#,
            total_count, completed_count, in_progress_count, pending_count, self
            .project_state.to_state_string(), self.context_window.to_context_string()
        )
    }
    pub fn get_next_chunk_to_work_on(&self) -> Option<String> {
        self.get_ready_chunks().first().cloned()
    }
    pub fn get_chunks_by_level(&self, level: ChunkingLevel) -> Vec<&CodeChunk> {
        self.chunks.values().filter(|c| c.level == level).collect()
    }
    pub fn get_dependency_chain(&self, chunk_id: &str) -> Vec<String> {
        let mut chain = Vec::new();
        let mut visited = HashSet::new();
        self._build_dependency_chain(chunk_id, &mut chain, &mut visited);
        chain
    }
    fn _build_dependency_chain(
        &self,
        chunk_id: &str,
        chain: &mut Vec<String>,
        visited: &mut HashSet<String>,
    ) {
        if visited.contains(chunk_id) {
            return;
        }
        visited.insert(chunk_id.to_string());
        if let Some(chunk) = self.chunks.get(chunk_id) {
            for dep in &chunk.dependencies {
                self._build_dependency_chain(dep, chain, visited);
            }
            chain.push(chunk_id.to_string());
        }
    }
}
pub fn parse_chunking_format(chunk_text: &str) -> Result<CodeChunk, String> {
    let start_tag = "<chunk>";
    let end_tag = "</chunk>";
    let start = chunk_text
        .find(start_tag)
        .ok_or_else(|| "Missing <chunk> start tag".to_string())?;
    let end = chunk_text
        .find(end_tag)
        .ok_or_else(|| "Missing </chunk> end tag".to_string())?;
    let content = &chunk_text[start + start_tag.len()..end];
    let parts: Vec<&str> = content.split(';').map(|s| s.trim()).collect();
    if parts.len() < 3 {
        return Err(
            "Invalid chunk format: need at least 3 parts (id; level; description)"
                .to_string(),
        );
    }
    let chunk_id = parts[0].to_string();
    let level = match parts[1].to_lowercase().as_str() {
        "project" => ChunkingLevel::Project,
        "module" => ChunkingLevel::Module,
        "class" => ChunkingLevel::Class,
        "method" => ChunkingLevel::Method,
        "block" => ChunkingLevel::Block,
        _ => return Err(format!("Invalid chunking level: {}", parts[1])),
    };
    let mut chunk = CodeChunk::new(chunk_id, level);
    if parts.len() > 3 {
        let dependencies: Vec<String> = parts[3]
            .split(',')
            .map(|s| s.trim().to_string())
            .collect();
        for dep in dependencies {
            if !dep.is_empty() {
                chunk.add_dependency(dep);
            }
        }
    }
    if parts.len() > 4 {
        chunk.set_code(parts[4].to_string());
    }
    Ok(chunk)
}
pub fn process_chunking_message(message: &str) -> Result<Vec<CodeChunk>, String> {
    let mut chunks = Vec::new();
    let chunk_pattern = r"<chunk>.*?</chunk>";
    let re = regex::Regex::new(chunk_pattern).unwrap();
    for mat in re.find_iter(message) {
        let chunk_text = mat.as_str();
        match parse_chunking_format(chunk_text) {
            Ok(chunk) => chunks.push(chunk),
            Err(e) => eprintln!("Warning: Failed to parse chunk: {}", e),
        }
    }
    Ok(chunks)
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_chunking_levels() {
        assert_eq!(ChunkingLevel::Project.max_tokens(), 100);
        assert_eq!(ChunkingLevel::Module.max_tokens(), 500);
        assert_eq!(ChunkingLevel::Class.max_tokens(), 1000);
        assert_eq!(ChunkingLevel::Method.max_tokens(), 300);
        assert_eq!(ChunkingLevel::Block.max_tokens(), 100);
    }
    #[test]
    fn test_project_state() {
        let mut state = ProjectState::new(1000);
        state.add_completed_module("module1".to_string());
        state.add_pending_module("module2".to_string());
        state.set_global_variable("API_KEY".to_string(), "secret123".to_string());
        assert_eq!(state.completed_modules.len(), 1);
        assert_eq!(state.pending_modules.len(), 1);
        assert_eq!(state.global_variables.len(), 1);
    }
    #[test]
    fn test_code_generation_graph() {
        let mut graph = CodeGenerationGraph::new(1000);
        graph.add_chunk("chunk1".to_string(), ChunkingLevel::Module, vec![]);
        graph
            .add_chunk(
                "chunk2".to_string(),
                ChunkingLevel::Class,
                vec!["chunk1".to_string()],
            );
        assert_eq!(graph.chunks.len(), 2);
        assert_eq!(graph.get_ready_chunks(), vec!["chunk1"]);
    }
    #[test]
    fn test_parse_chunking_format() {
        let chunk_text = "<chunk>chunk1; module; Create database handler; chunk0; import sqlite3</chunk>";
        let chunk = parse_chunking_format(chunk_text).unwrap();
        assert_eq!(chunk.chunk_id, "chunk1");
        assert_eq!(chunk.level, ChunkingLevel::Module);
        assert_eq!(chunk.dependencies.len(), 1);
        assert_eq!(chunk.dependencies[0], "chunk0");
    }
}

File: tdz/py/chunking.py
"""Code generation chunking for Todozi.

This module handles breaking down large code generation tasks into manageable chunks
with dependency tracking and validation.
"""

from dataclasses import dataclass, field
from enum import Enum
from typing import Dict, List, Optional, Set
from datetime import datetime
import re


class ChunkingLevel(str, Enum):
    """Levels of code chunking."""
    PROJECT = "project"
    MODULE = "module"
    CLASS = "class"
    METHOD = "method"
    BLOCK = "block"

    def max_tokens(self) -> int:
        """Get maximum tokens for this level."""
        return {
            ChunkingLevel.PROJECT: 100,
            ChunkingLevel.MODULE: 500,
            ChunkingLevel.CLASS: 1000,
            ChunkingLevel.METHOD: 300,
            ChunkingLevel.BLOCK: 100,
        }[self]

    def description(self) -> str:
        """Get description for this level."""
        return {
            ChunkingLevel.PROJECT: "High-level project planning and architecture",
            ChunkingLevel.MODULE: "Major system components and interfaces",
            ChunkingLevel.CLASS: "Class definitions and major functions",
            ChunkingLevel.METHOD: "Individual methods and helper functions",
            ChunkingLevel.BLOCK: "Small code blocks and error handling",
        }[self]

    def example(self) -> str:
        """Get example for this level."""
        return {
            ChunkingLevel.PROJECT: "Build web scraper with database storage",
            ChunkingLevel.MODULE: "Create database handler module",
            ChunkingLevel.CLASS: "Implement DatabaseConnection class",
            ChunkingLevel.METHOD: "Write insert_record method",
            ChunkingLevel.BLOCK: "Add error handling for connection timeout",
        }[self]


class ChunkStatus(str, Enum):
    """Status of a code chunk."""
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    VALIDATED = "validated"
    FAILED = "failed"


@dataclass
class ProjectState:
    """State of the overall project generation."""
    total_lines: int = 0
    max_lines: int = 1000
    current_module: str = ""
    dependencies: List[str] = field(default_factory=list)
    completed_modules: List[str] = field(default_factory=list)
    pending_modules: List[str] = field(default_factory=list)
    global_variables: Dict[str, str] = field(default_factory=dict)
    created_at: datetime = field(default_factory=lambda: datetime.utcnow())
    updated_at: datetime = field(default_factory=lambda: datetime.utcnow())

    def to_state_string(self) -> str:
        """Convert to formatted state string."""
        return f"""<project_state>
- Total lines written: {self.total_lines}/{self.max_lines}
- Current module: {self.current_module}
- Dependencies: {', '.join(self.dependencies)}
- Completed modules: {', '.join(self.completed_modules)}
- Pending modules: {', '.join(self.pending_modules)}
- Global variables: {', '.join(f'{k}={v}' for k, v in self.global_variables.items())}
- Created: {self.created_at.strftime('%Y-%m-%d %H:%M:%S')}
- Updated: {self.updated_at.strftime('%Y-%m-%d %H:%M:%S')}
</project_state>"""

    def add_completed_module(self, module: str):
        """Add a completed module."""
        if module not in self.completed_modules:
            self.completed_modules.append(module)
            self.updated_at = datetime.utcnow()

    def add_pending_module(self, module: str):
        """Add a pending module."""
        if module not in self.pending_modules:
            self.pending_modules.append(module)
            self.updated_at = datetime.utcnow()

    def set_global_variable(self, key: str, value: str):
        """Set a global variable."""
        self.global_variables[key] = value
        self.updated_at = datetime.utcnow()

    def increment_lines(self, lines: int):
        """Increment line count."""
        self.total_lines += lines
        self.updated_at = datetime.utcnow()


@dataclass
class ContextWindow:
    """Context window for tracking code generation state."""
    previous_class: str = ""
    current_class: str = ""
    next_planned: str = ""
    global_vars_in_scope: List[str] = field(default_factory=list)
    imports_used: List[str] = field(default_factory=list)
    function_signatures: Dict[str, str] = field(default_factory=dict)
    error_patterns_seen: List[str] = field(default_factory=list)
    created_at: datetime = field(default_factory=lambda: datetime.utcnow())
    updated_at: datetime = field(default_factory=lambda: datetime.utcnow())

    def to_context_string(self) -> str:
        """Convert to formatted context string."""
        return f"""<context_window>
- Previous class: {self.previous_class}
- Current class: {self.current_class}
- Next planned: {self.next_planned}
- Global variables in scope: {', '.join(self.global_vars_in_scope)}
- Imports used: {', '.join(self.imports_used)}
- Function signatures: {', '.join(f'{k}: {v}' for k, v in self.function_signatures.items())}
- Error patterns seen: {', '.join(self.error_patterns_seen)}
- Created: {self.created_at.strftime('%Y-%m-%d %H:%M:%S')}
- Updated: {self.updated_at.strftime('%Y-%m-%d %H:%M:%S')}
</context_window>"""

    def add_import(self, import_stmt: str):
        """Add an import."""
        if import_stmt not in self.imports_used:
            self.imports_used.append(import_stmt)
            self.updated_at = datetime.utcnow()

    def add_function_signature(self, name: str, signature: str):
        """Add a function signature."""
        self.function_signatures[name] = signature
        self.updated_at = datetime.utcnow()

    def add_error_pattern(self, pattern: str):
        """Add an error pattern."""
        if pattern not in self.error_patterns_seen:
            self.error_patterns_seen.append(pattern)
            self.updated_at = datetime.utcnow()

    def set_current_class(self, class_name: str):
        """Set the current class being worked on."""
        self.previous_class = self.current_class
        self.current_class = class_name
        self.updated_at = datetime.utcnow()


@dataclass
class CodeChunk:
    """A chunk of code to be generated."""
    chunk_id: str
    status: ChunkStatus = ChunkStatus.PENDING
    dependencies: List[str] = field(default_factory=list)
    code: str = ""
    tests: str = ""
    validated: bool = False
    level: ChunkingLevel = ChunkingLevel.METHOD
    estimated_tokens: int = 0
    created_at: datetime = field(default_factory=lambda: datetime.utcnow())
    updated_at: datetime = field(default_factory=lambda: datetime.utcnow())

    @staticmethod
    def new(chunk_id: str, level: ChunkingLevel) -> 'CodeChunk':
        """Create a new code chunk."""
        return CodeChunk(chunk_id=chunk_id, level=level)

    def add_dependency(self, dep: str):
        """Add a dependency."""
        if dep not in self.dependencies:
            self.dependencies.append(dep)
            self.updated_at = datetime.utcnow()

    def set_code(self, code: str):
        """Set the code for this chunk."""
        self.estimated_tokens = len(code.split())
        self.code = code
        self.updated_at = datetime.utcnow()

    def set_tests(self, tests: str):
        """Set the tests for this chunk."""
        self.tests = tests
        self.updated_at = datetime.utcnow()

    def mark_completed(self):
        """Mark the chunk as completed."""
        self.status = ChunkStatus.COMPLETED
        self.updated_at = datetime.utcnow()

    def mark_validated(self):
        """Mark the chunk as validated."""
        self.validated = True
        self.status = ChunkStatus.VALIDATED
        self.updated_at = datetime.utcnow()

    def mark_failed(self):
        """Mark the chunk as failed."""
        self.status = ChunkStatus.FAILED
        self.updated_at = datetime.utcnow()


class CodeGenerationGraph:
    """Graph managing code generation chunks and dependencies."""

    def __init__(self, max_lines: int):
        self.chunks: Dict[str, CodeChunk] = {}
        self.project_state = ProjectState(max_lines=max_lines)
        self.context_window = ContextWindow()

    def add_chunk(self, chunk_id: str, level: ChunkingLevel, deps: List[str]):
        """Add a new chunk to the graph."""
        chunk = CodeChunk.new(chunk_id, level)
        for dep in deps:
            chunk.add_dependency(dep)
        self.chunks[chunk_id] = chunk

    def get_ready_chunks(self) -> List[str]:
        """Get chunks that are ready to be worked on."""
        ready = []
        for chunk_id, chunk in self.chunks.items():
            if chunk.status == ChunkStatus.PENDING:
                deps_satisfied = all(
                    self.chunks.get(dep) and
                    self.chunks[dep].status in [ChunkStatus.COMPLETED, ChunkStatus.VALIDATED]
                    for dep in chunk.dependencies
                )
                if deps_satisfied:
                    ready.append(chunk_id)
        return ready

    def get_chunk(self, chunk_id: str) -> Optional[CodeChunk]:
        """Get a chunk by ID."""
        return self.chunks.get(chunk_id)

    def get_chunk_mut(self, chunk_id: str) -> Optional[CodeChunk]:
        """Get a mutable reference to a chunk."""
        return self.chunks.get(chunk_id)

    def update_chunk_code(self, chunk_id: str, code: str):
        """Update the code for a chunk."""
        chunk = self.chunks.get(chunk_id)
        if not chunk:
            raise ValueError(f"Chunk {chunk_id} not found")

        chunk.set_code(code)
        self.project_state.increment_lines(len(code.splitlines()))

    def update_chunk_tests(self, chunk_id: str, tests: str):
        """Update the tests for a chunk."""
        chunk = self.chunks.get(chunk_id)
        if not chunk:
            raise ValueError(f"Chunk {chunk_id} not found")

        chunk.set_tests(tests)

    def mark_chunk_completed(self, chunk_id: str):
        """Mark a chunk as completed."""
        chunk = self.chunks.get(chunk_id)
        if not chunk:
            raise ValueError(f"Chunk {chunk_id} not found")

        chunk.mark_completed()
        self.project_state.add_completed_module(chunk_id)

    def mark_chunk_validated(self, chunk_id: str):
        """Mark a chunk as validated."""
        chunk = self.chunks.get(chunk_id)
        if not chunk:
            raise ValueError(f"Chunk {chunk_id} not found")

        chunk.mark_validated()

    def get_project_summary(self) -> str:
        """Get a summary of the project state."""
        completed_count = sum(
            1 for c in self.chunks.values()
            if c.status in [ChunkStatus.COMPLETED, ChunkStatus.VALIDATED]
        )
        total_count = len(self.chunks)
        pending_count = sum(
            1 for c in self.chunks.values()
            if c.status == ChunkStatus.PENDING
        )
        in_progress_count = sum(
            1 for c in self.chunks.values()
            if c.status == ChunkStatus.IN_PROGRESS
        )

        return f"""<project_summary>
- Total chunks: {total_count}
- Completed: {completed_count}
- In progress: {in_progress_count}
- Pending: {pending_count}
- Project state: {self.project_state.to_state_string()}
- Context window: {self.context_window.to_context_string()}
</project_summary>"""

    def get_next_chunk_to_work_on(self) -> Optional[str]:
        """Get the next chunk to work on."""
        ready = self.get_ready_chunks()
        return ready[0] if ready else None

    def get_chunks_by_level(self, level: ChunkingLevel) -> List[CodeChunk]:
        """Get all chunks at a specific level."""
        return [c for c in self.chunks.values() if c.level == level]

    def get_dependency_chain(self, chunk_id: str) -> List[str]:
        """Get the full dependency chain for a chunk."""
        chain = []
        visited: Set[str] = set()
        self._build_dependency_chain(chunk_id, chain, visited)
        return chain

    def _build_dependency_chain(self, chunk_id: str, chain: List[str], visited: Set[str]):
        """Recursively build dependency chain."""
        if chunk_id in visited:
            return

        visited.add(chunk_id)
        chunk = self.chunks.get(chunk_id)
        if chunk:
            for dep in chunk.dependencies:
                self._build_dependency_chain(dep, chain, visited)
            chain.append(chunk_id)


def parse_chunking_format(chunk_text: str) -> CodeChunk:
    """Parse a chunk from XML-like format."""
    start_tag = "<chunk>"
    end_tag = "</chunk>"

    start = chunk_text.find(start_tag)
    if start == -1:
        raise ValueError("Missing <chunk> start tag")

    end = chunk_text.find(end_tag)
    if end == -1:
        raise ValueError("Missing </chunk> end tag")

    content = chunk_text[start + len(start_tag):end]
    parts = [s.strip() for s in content.split(';')]

    if len(parts) < 3:
        raise ValueError("Invalid chunk format: need at least 3 parts (id; level; description)")

    chunk_id = parts[0]

    # Parse level
    level_str = parts[1].lower()
    try:
        level = ChunkingLevel(level_str)
    except ValueError:
        raise ValueError(f"Invalid chunking level: {parts[1]}")

    chunk = CodeChunk.new(chunk_id, level)

    # Parse dependencies
    if len(parts) > 3:
        dependencies = [s.strip() for s in parts[3].split(',') if s.strip()]
        for dep in dependencies:
            chunk.add_dependency(dep)

    # Parse code
    if len(parts) > 4:
        chunk.set_code(parts[4])

    return chunk


def process_chunking_message(message: str) -> List[CodeChunk]:
    """Process a message and extract all chunks."""
    chunks = []
    chunk_pattern = r"<chunk>.*?</chunk>"

    for match in re.finditer(chunk_pattern, message, re.DOTALL):
        chunk_text = match.group(0)
        try:
            chunk = parse_chunking_format(chunk_text)
            chunks.append(chunk)
        except ValueError as e:
            print(f"Warning: Failed to parse chunk: {e}")

    return chunks
